cmd/arc_summary/arc_summary3.py

```diff
@@ -0,0 +1,849 @@
+    Division by zero is handled by returning 'n/a'; no error
+    is called.
+    """
+
+    v1 = float(value1)
+    v2 = float(value2)
+
+    try:
+        perc = 100 * v1/v2
+    except ZeroDivisionError:
+        result = 'n/a'
+    else:
+        result = '{0:0.1f} %'.format(perc)
+
+    if result == '0.0 %' and v1 > 0:
+        result = '< 0.1 %'
+
+    return result
+
+
+def format_raw_line(name, value):
+    """For the --raw option for the tunable and SPL outputs, decide on the
+    correct formatting based on the --alternate flag.
+    """
+
+    if ARGS.alt:
+        result = '{0}{1}={2}'.format(INDENT, name, value)
+    else:
+        spc = LINE_LENGTH-(len(INDENT)+len(value))
+        result = '{0}{1:<{spc}}{2}'.format(INDENT, name, value, spc=spc)
+
+    return result
+
+
+def get_kstats():
+    """Collect information on the ZFS subsystem from the /proc Linux virtual
+    file system. The step does not perform any further processing, giving us
+    the option to only work on what is actually needed. The name "kstat" is a
+    holdover from the Solaris utility of the same name.
+    """
+
+    result = {}
+    secs = SECTION_PATHS.values()
+
+    for section in secs:
+
+        with open(PROC_PATH+section, 'r') as proc_location:
+            lines = [line for line in proc_location]
+
+        del lines[0:2]  # Get rid of header
+        result[section] = lines
+
+    return result
+
+
+def get_spl_tunables(PATH):
+    """Collect information on the Solaris Porting Layer (SPL) or the
+    tunables, depending on the PATH given. Does not check if PATH is
+    legal.
+    """
+
+    result = {}
+    parameters = os.listdir(PATH)
+
+    for name in parameters:
+
+        with open(PATH+name, 'r') as para_file:
+            value = para_file.read()
+            result[name] = value.strip()
+
+    return result
+
+
+def get_descriptions(request):
+    """Get the decriptions of the Solaris Porting Layer (SPL) or the
+    tunables, return with minimal formatting.
+    """
+
+    if request not in ('spl', 'zfs'):
+        print('ERROR: description of "{0}" requested)'.format(request))
+        sys.exit(1)
+
+    descs = {}
+    target_prefix = 'parm:'
+
+    # We would prefer to do this with /sys/modules -- see the discusion at
+    # get_version() -- but there isn't a way to get the descriptions from
+    # there, so we fall back on modinfo
+    command = ["/sbin/modinfo", request, "-0"]
+
+    # The recommended way to do this is with subprocess.run(). However,
+    # some installed versions of Python are < 3.5, so we offer them
+    # the option of doing it the old way (for now)
+    info = ''
+    if 'run' in dir(subprocess):
+        info = subprocess.run(command, stdout=subprocess.PIPE,
+                              universal_newlines=True)
+        raw_output = info.stdout.split('\0')
+    else:
+        info = subprocess.check_output(command, universal_newlines=True)
+        raw_output = info.split('\0')
```

Comment:
Almost there!  Everything now works as intended when the kmods exist. However we still throw an ugly exception when they're missing.  Can you just handle this a bit more gracefully and not print an error.  This case can legitimately come up for at least two reasons.

* ZFS was built direclty in to the kernel and not as an independant kmod, or
* a developer is using a custom kmod which was built in the zfs srouce in-tree and not loaded with modprobe


```
modinfo: ERROR: Module spl not found.
Traceback (most recent call last):
  File "./cmd/arc_summary/arc_summary3.py", line 849, in <module>
    main()
  File "./cmd/arc_summary/arc_summary3.py", line 843, in main
    section_calls[section](kstats)
  File "./cmd/arc_summary/arc_summary3.py", line 693, in section_spl
    descriptions = get_descriptions('spl')
  File "./cmd/arc_summary/arc_summary3.py", line 305, in get_descriptions
    info = subprocess.check_output(command, universal_newlines=True)
  File "/usr/lib64/python3.4/subprocess.py", line 617, in check_output
    raise CalledProcessError(retcode, process.args, output=output)
subprocess.CalledProcessError: Command '['/sbin/modinfo', 'spl', '-0']' returned non-zero exit status 1
```